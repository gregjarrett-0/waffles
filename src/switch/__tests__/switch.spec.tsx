import { useRef, useEffect } from 'react';
import { render, fireEvent } from '@testing-library/react';

import { Switch } from '../index';

const MOCKED_ID = '123abC';

jest.mock('nanoid', () => {
  return {
    nanoid: () => MOCKED_ID,
  };
});

function TestRefSwitch() {
  const inputRef = useRef<HTMLInputElement>(null);

  useEffect(() => {
    inputRef.current?.focus();
  }, []);

  return (
    <Switch ref={inputRef} onChange={jest.fn()}>
      Test label
    </Switch>
  );
}

describe('Switch', () => {
  it('renders a label and sets correct aria attributes when not checked', () => {
    const { getByRole, getByText } = render(
      <Switch onChange={jest.fn()}>Dark mode</Switch>,
    );

    const input = getByRole('switch');
    const label = getByText('Dark mode');

    expect(input).toHaveAttribute('aria-checked', 'false');
    expect(label).toBeInTheDocument();
  });

  it('renders a label and sets correct aria attributes when checked', () => {
    const { getByRole, getByText } = render(
      <Switch checked onChange={jest.fn()}>
        Dark mode
      </Switch>,
    );

    const input = getByRole('switch');
    const label = getByText('Dark mode');

    expect(input).toHaveAttribute('aria-checked', 'true');
    expect(label).toBeInTheDocument();
  });

  it('handles click event correctly', () => {
    const handleClick = jest.fn();
    const { getByText } = render(
      <Switch onChange={handleClick}>Dark mode</Switch>,
    );

    const label = getByText('Dark mode');
    fireEvent.click(label);

    expect(handleClick).toHaveBeenCalledTimes(1);
  });

  it('handles focus correctly', () => {
    const handleFocus = jest.fn();
    const { getByLabelText } = render(
      <Switch onFocus={handleFocus} onChange={jest.fn()}>
        Dark mode
      </Switch>,
    );

    const input = getByLabelText('Dark mode', { selector: 'input' });
    fireEvent.focus(input);

    expect(handleFocus).toHaveBeenCalledTimes(1);
  });

  it('sets the data attribute', () => {
    const { getByTestId } = render(
      <Switch data-testid="dark-mode-switch" onChange={jest.fn()}>
        Dark mode
      </Switch>,
    );

    const input = getByTestId('dark-mode-switch');

    expect(input).toBeInTheDocument();
  });

  it('sets correct aria attributes when has error', () => {
    const { getByLabelText } = render(
      <Switch error onChange={jest.fn()}>
        Dark mode
      </Switch>,
    );

    const input = getByLabelText('Dark mode', { selector: 'input' });

    expect(input).toHaveAttribute('aria-invalid', 'true');
  });

  it('renders the disabled state', () => {
    const { getByLabelText } = render(
      <Switch disabled onChange={jest.fn()}>
        Dark mode
      </Switch>,
    );

    const input = getByLabelText('Dark mode', { selector: 'input' });

    expect(input).toBeDisabled();
  });

  it('accepts ref and could be focused programmatically', () => {
    const { container } = render(<TestRefSwitch />);

    const input = container.querySelector('input');

    expect(input).toHaveFocus();
  });

  it('renders basic snapshot', () => {
    const { container } = render(
      <Switch onChange={jest.fn()}>Dark mode</Switch>,
    );

    const switchComponent = container.firstChild;

    expect(switchComponent).toMatchSnapshot();
  });

  it('renders snapshot of switch without label', () => {
    const { container } = render(
      <Switch aria-label="Switch no label" onChange={jest.fn()} />,
    );

    const switchComponent = container.firstChild;

    expect(switchComponent).toMatchSnapshot();
  });

  it('renders snapshot of toggled state', () => {
    const { container } = render(
      <Switch checked onChange={jest.fn()}>
        Dark mode
      </Switch>,
    );

    const switchComponent = container.firstChild;

    expect(switchComponent).toMatchSnapshot();
  });

  it('renders snapshot with error', () => {
    const { container } = render(
      <Switch error onChange={jest.fn()}>
        Dark mode
      </Switch>,
    );

    const switchComponent = container.firstChild;

    expect(switchComponent).toMatchSnapshot();
  });

  it('renders snapshot of inverted', () => {
    const { container } = render(
      <Switch inverted onChange={jest.fn()}>
        Dark mode
      </Switch>,
    );

    const switchComponent = container.firstChild;

    expect(switchComponent).toMatchSnapshot();
  });

  it('renders snapshot of disabled state', () => {
    const { container } = render(
      <Switch disabled onChange={jest.fn()}>
        Dark mode
      </Switch>,
    );

    const switchComponent = container.firstChild;

    expect(switchComponent).toMatchSnapshot();
  });

  it('renders snapshot of focused state', () => {
    const { container, getByLabelText } = render(
      <Switch onChange={jest.fn()}>Dark mode</Switch>,
    );

    const input = getByLabelText('Dark mode', { selector: 'input' });
    fireEvent.focus(input);
    const switchComponent = container.firstChild;

    expect(switchComponent).toMatchSnapshot();
  });
});
