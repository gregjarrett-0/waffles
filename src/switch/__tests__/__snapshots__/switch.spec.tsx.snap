// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Switch renders basic snapshot 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  opacity: 1;
}

.emotion-1 {
  color: #05192D;
  font-family: Studio-Feixen-Sans,Arial,sans-serif;
  font-size: 14px;
  font-weight: 400;
  line-height: 1;
  margin: 0;
  padding: 0;
  line-height: 1.25;
  padding-top: 3px;
  padding-right: 8px;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  color: #05192D;
  cursor: pointer;
}

.emotion-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: start;
  -ms-flex-pack: start;
  -webkit-justify-content: flex-start;
  justify-content: flex-start;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  position: relative;
  height: 22px;
  width: 36px;
  padding: 2px;
  border-radius: 12px;
  -webkit-transition: background-color 75ms ease-out;
  transition: background-color 75ms ease-out;
  background-color: #848492;
}

.emotion-2::after {
  content: '';
  display: block;
  position: absolute;
  z-index: 1;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  margin: -2px;
  border-radius: 14px;
  pointer-events: none;
  -webkit-transition: box-shadow 125ms ease-out;
  transition: box-shadow 125ms ease-out;
}

.emotion-3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  color: #009BD8;
  background-color: #FFFFFF;
  height: 18px;
  width: 18px;
  border-radius: 50%;
  -webkit-transform: translateX(0px);
  -moz-transform: translateX(0px);
  -ms-transform: translateX(0px);
  transform: translateX(0px);
  -webkit-transition: -webkit-transform 75ms ease-out;
  transition: transform 75ms ease-out;
}

.emotion-4 {
  position: absolute;
  opacity: 0;
  height: 22px;
  width: 36px;
  margin: 0;
  padding: 0;
  overflow: hidden;
  cursor: pointer;
}

<div
  class="emotion-0"
>
  <label
    class="emotion-1"
    for="switch-123abC"
  >
    Dark mode
  </label>
  <div
    class="emotion-2"
  >
    <span
      class="emotion-3"
    />
    <input
      aria-checked="false"
      aria-invalid="false"
      class="emotion-4"
      id="switch-123abC"
      role="switch"
      type="checkbox"
    />
  </div>
</div>
`;

exports[`Switch renders snapshot of disabled state 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  opacity: 0.6;
}

.emotion-1 {
  color: #05192D;
  font-family: Studio-Feixen-Sans,Arial,sans-serif;
  font-size: 14px;
  font-weight: 400;
  line-height: 1;
  margin: 0;
  padding: 0;
  line-height: 1.25;
  padding-top: 3px;
  padding-right: 8px;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  color: #05192D;
  cursor: default;
}

.emotion-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: start;
  -ms-flex-pack: start;
  -webkit-justify-content: flex-start;
  justify-content: flex-start;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  position: relative;
  height: 22px;
  width: 36px;
  padding: 2px;
  border-radius: 12px;
  -webkit-transition: background-color 75ms ease-out;
  transition: background-color 75ms ease-out;
  background-color: #848492;
}

.emotion-2::after {
  content: '';
  display: block;
  position: absolute;
  z-index: 1;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  margin: -2px;
  border-radius: 14px;
  pointer-events: none;
  -webkit-transition: box-shadow 125ms ease-out;
  transition: box-shadow 125ms ease-out;
}

.emotion-3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  color: #009BD8;
  background-color: #FFFFFF;
  height: 18px;
  width: 18px;
  border-radius: 50%;
  -webkit-transform: translateX(0px);
  -moz-transform: translateX(0px);
  -ms-transform: translateX(0px);
  transform: translateX(0px);
  -webkit-transition: -webkit-transform 75ms ease-out;
  transition: transform 75ms ease-out;
}

.emotion-4 {
  position: absolute;
  opacity: 0;
  height: 22px;
  width: 36px;
  margin: 0;
  padding: 0;
  overflow: hidden;
  cursor: default;
}

<div
  class="emotion-0"
>
  <label
    class="emotion-1"
    for="switch-123abC"
  >
    Dark mode
  </label>
  <div
    class="emotion-2"
  >
    <span
      class="emotion-3"
    />
    <input
      aria-checked="false"
      aria-invalid="false"
      class="emotion-4"
      disabled=""
      id="switch-123abC"
      role="switch"
      type="checkbox"
    />
  </div>
</div>
`;

exports[`Switch renders snapshot of focused state 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  opacity: 1;
}

.emotion-1 {
  color: #05192D;
  font-family: Studio-Feixen-Sans,Arial,sans-serif;
  font-size: 14px;
  font-weight: 400;
  line-height: 1;
  margin: 0;
  padding: 0;
  line-height: 1.25;
  padding-top: 3px;
  padding-right: 8px;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  color: #05192D;
  cursor: pointer;
}

.emotion-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: start;
  -ms-flex-pack: start;
  -webkit-justify-content: flex-start;
  justify-content: flex-start;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  position: relative;
  height: 22px;
  width: 36px;
  padding: 2px;
  border-radius: 12px;
  -webkit-transition: background-color 75ms ease-out;
  transition: background-color 75ms ease-out;
  background-color: #848492;
}

.emotion-2::after {
  content: '';
  display: block;
  position: absolute;
  z-index: 1;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  margin: -2px;
  border-radius: 14px;
  pointer-events: none;
  -webkit-transition: box-shadow 125ms ease-out;
  transition: box-shadow 125ms ease-out;
}

.emotion-2::after {
  box-shadow: 0 0 0 2px #009BD8;
}

.emotion-3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  color: #009BD8;
  background-color: #FFFFFF;
  height: 18px;
  width: 18px;
  border-radius: 50%;
  -webkit-transform: translateX(0px);
  -moz-transform: translateX(0px);
  -ms-transform: translateX(0px);
  transform: translateX(0px);
  -webkit-transition: -webkit-transform 75ms ease-out;
  transition: transform 75ms ease-out;
}

.emotion-4 {
  position: absolute;
  opacity: 0;
  height: 22px;
  width: 36px;
  margin: 0;
  padding: 0;
  overflow: hidden;
  cursor: pointer;
}

<div
  class="emotion-0"
>
  <label
    class="emotion-1"
    for="switch-123abC"
  >
    Dark mode
  </label>
  <div
    class="emotion-2"
  >
    <span
      class="emotion-3"
    />
    <input
      aria-checked="false"
      aria-invalid="false"
      class="emotion-4"
      id="switch-123abC"
      role="switch"
      type="checkbox"
    />
  </div>
</div>
`;

exports[`Switch renders snapshot of inverted 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  opacity: 1;
}

.emotion-1 {
  color: #05192D;
  font-family: Studio-Feixen-Sans,Arial,sans-serif;
  font-size: 14px;
  font-weight: 400;
  line-height: 1;
  margin: 0;
  padding: 0;
  line-height: 1.25;
  padding-top: 3px;
  padding-right: 8px;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  color: #FFFFFF;
  cursor: pointer;
}

.emotion-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: start;
  -ms-flex-pack: start;
  -webkit-justify-content: flex-start;
  justify-content: flex-start;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  position: relative;
  height: 22px;
  width: 36px;
  padding: 2px;
  border-radius: 12px;
  -webkit-transition: background-color 75ms ease-out;
  transition: background-color 75ms ease-out;
  background-color: #848492;
}

.emotion-2::after {
  content: '';
  display: block;
  position: absolute;
  z-index: 1;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  margin: -2px;
  border-radius: 14px;
  pointer-events: none;
  -webkit-transition: box-shadow 125ms ease-out;
  transition: box-shadow 125ms ease-out;
}

.emotion-3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  color: #06BDFC;
  background-color: #05192D;
  height: 18px;
  width: 18px;
  border-radius: 50%;
  -webkit-transform: translateX(0px);
  -moz-transform: translateX(0px);
  -ms-transform: translateX(0px);
  transform: translateX(0px);
  -webkit-transition: -webkit-transform 75ms ease-out;
  transition: transform 75ms ease-out;
}

.emotion-4 {
  position: absolute;
  opacity: 0;
  height: 22px;
  width: 36px;
  margin: 0;
  padding: 0;
  overflow: hidden;
  cursor: pointer;
}

<div
  class="emotion-0"
>
  <label
    class="emotion-1"
    for="switch-123abC"
  >
    Dark mode
  </label>
  <div
    class="emotion-2"
  >
    <span
      class="emotion-3"
    />
    <input
      aria-checked="false"
      aria-invalid="false"
      class="emotion-4"
      id="switch-123abC"
      role="switch"
      type="checkbox"
    />
  </div>
</div>
`;

exports[`Switch renders snapshot of switch without label 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  opacity: 1;
}

.emotion-1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: start;
  -ms-flex-pack: start;
  -webkit-justify-content: flex-start;
  justify-content: flex-start;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  position: relative;
  height: 22px;
  width: 36px;
  padding: 2px;
  border-radius: 12px;
  -webkit-transition: background-color 75ms ease-out;
  transition: background-color 75ms ease-out;
  background-color: #848492;
}

.emotion-1::after {
  content: '';
  display: block;
  position: absolute;
  z-index: 1;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  margin: -2px;
  border-radius: 14px;
  pointer-events: none;
  -webkit-transition: box-shadow 125ms ease-out;
  transition: box-shadow 125ms ease-out;
}

.emotion-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  color: #009BD8;
  background-color: #FFFFFF;
  height: 18px;
  width: 18px;
  border-radius: 50%;
  -webkit-transform: translateX(0px);
  -moz-transform: translateX(0px);
  -ms-transform: translateX(0px);
  transform: translateX(0px);
  -webkit-transition: -webkit-transform 75ms ease-out;
  transition: transform 75ms ease-out;
}

.emotion-3 {
  position: absolute;
  opacity: 0;
  height: 22px;
  width: 36px;
  margin: 0;
  padding: 0;
  overflow: hidden;
  cursor: pointer;
}

<div
  class="emotion-0"
>
  <div
    class="emotion-1"
  >
    <span
      class="emotion-2"
    />
    <input
      aria-checked="false"
      aria-invalid="false"
      aria-label="Switch no label"
      class="emotion-3"
      role="switch"
      type="checkbox"
    />
  </div>
</div>
`;

exports[`Switch renders snapshot of toggled state 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  opacity: 1;
}

.emotion-1 {
  color: #05192D;
  font-family: Studio-Feixen-Sans,Arial,sans-serif;
  font-size: 14px;
  font-weight: 400;
  line-height: 1;
  margin: 0;
  padding: 0;
  line-height: 1.25;
  padding-top: 3px;
  padding-right: 8px;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  color: #05192D;
  cursor: pointer;
}

.emotion-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: start;
  -ms-flex-pack: start;
  -webkit-justify-content: flex-start;
  justify-content: flex-start;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  position: relative;
  height: 22px;
  width: 36px;
  padding: 2px;
  border-radius: 12px;
  -webkit-transition: background-color 75ms ease-out;
  transition: background-color 75ms ease-out;
  background-color: #009BD8;
}

.emotion-2::after {
  content: '';
  display: block;
  position: absolute;
  z-index: 1;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  margin: -2px;
  border-radius: 14px;
  pointer-events: none;
  -webkit-transition: box-shadow 125ms ease-out;
  transition: box-shadow 125ms ease-out;
}

.emotion-3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  color: #009BD8;
  background-color: #FFFFFF;
  height: 18px;
  width: 18px;
  border-radius: 50%;
  -webkit-transform: translateX(14px);
  -moz-transform: translateX(14px);
  -ms-transform: translateX(14px);
  transform: translateX(14px);
  -webkit-transition: -webkit-transform 75ms ease-out;
  transition: transform 75ms ease-out;
}

.emotion-4 {
  position: absolute;
  opacity: 0;
  height: 22px;
  width: 36px;
  margin: 0;
  padding: 0;
  overflow: hidden;
  cursor: pointer;
}

<div
  class="emotion-0"
>
  <label
    class="emotion-1"
    for="switch-123abC"
  >
    Dark mode
  </label>
  <div
    class="emotion-2"
  >
    <span
      class="emotion-3"
    >
      <svg
        aria-hidden="true"
        height="14"
        viewBox="0 0 18 18"
        width="14"
      >
        <path
          d="M13.746 4.337a1.015 1.015 0 0 1 1.409-.099c.417.354.462.97.101 1.378l-7.13 8.047a1.015 1.015 0 0 1-1.483.03L2.771 9.67a.961.961 0 0 1 .044-1.38 1.015 1.015 0 0 1 1.412.041l3.113 3.235 6.406-7.229Z"
          fill="currentColor"
        />
      </svg>
    </span>
    <input
      aria-checked="true"
      aria-invalid="false"
      checked=""
      class="emotion-4"
      id="switch-123abC"
      role="switch"
      type="checkbox"
    />
  </div>
</div>
`;

exports[`Switch renders snapshot with error 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  opacity: 1;
}

.emotion-1 {
  color: #05192D;
  font-family: Studio-Feixen-Sans,Arial,sans-serif;
  font-size: 14px;
  font-weight: 400;
  line-height: 1;
  margin: 0;
  padding: 0;
  line-height: 1.25;
  padding-top: 3px;
  padding-right: 8px;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  color: #05192D;
  cursor: pointer;
}

.emotion-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: start;
  -ms-flex-pack: start;
  -webkit-justify-content: flex-start;
  justify-content: flex-start;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  position: relative;
  height: 22px;
  width: 36px;
  padding: 2px;
  border-radius: 12px;
  -webkit-transition: background-color 75ms ease-out;
  transition: background-color 75ms ease-out;
  background-color: #848492;
  background-color: #DD3400;
}

.emotion-2::after {
  content: '';
  display: block;
  position: absolute;
  z-index: 1;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  margin: -2px;
  border-radius: 14px;
  pointer-events: none;
  -webkit-transition: box-shadow 125ms ease-out;
  transition: box-shadow 125ms ease-out;
}

.emotion-3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  color: #009BD8;
  background-color: #FFFFFF;
  height: 18px;
  width: 18px;
  border-radius: 50%;
  -webkit-transform: translateX(0px);
  -moz-transform: translateX(0px);
  -ms-transform: translateX(0px);
  transform: translateX(0px);
  -webkit-transition: -webkit-transform 75ms ease-out;
  transition: transform 75ms ease-out;
}

.emotion-4 {
  position: absolute;
  opacity: 0;
  height: 22px;
  width: 36px;
  margin: 0;
  padding: 0;
  overflow: hidden;
  cursor: pointer;
}

<div
  class="emotion-0"
>
  <label
    class="emotion-1"
    for="switch-123abC"
  >
    Dark mode
  </label>
  <div
    class="emotion-2"
  >
    <span
      class="emotion-3"
    />
    <input
      aria-checked="false"
      aria-invalid="true"
      class="emotion-4"
      id="switch-123abC"
      role="switch"
      type="checkbox"
    />
  </div>
</div>
`;
